src/util/misc.rs

```diff
@@ -17,11 +17,188 @@
+                compressed: compressed,
+            }
+        }
+
+        /// Serialize to bytes.
+        pub fn serialize(&self) -> [u8; 65] {
+            let (recid, raw) = self.signature.serialize_compact();
+            let mut serialized = [0u8; 65];
+            serialized[0] = 27;
+            serialized[0] += recid.to_i32() as u8;
+            if self.compressed {
+                serialized[0] += 4;
+            }
+            serialized[1..].copy_from_slice(&raw[..]);
+            serialized
+        }
+
+        /// Create from a byte slice.
+        pub fn from_slice(bytes: &[u8]) -> Result<MessageSignature, MessageSignatureError> {
+            if bytes.len() != 65 {
+                return Err(MessageSignatureError::InvalidLength);
+            }
+            // We just check this here so we can safely subtract further.
+            if bytes[0] < 27 {
+                return Err(MessageSignatureError::InvalidEncoding(secp256k1::Error::InvalidRecoveryId));
+            };
+            let recid = RecoveryId::from_i32(((bytes[0] - 27) & 0x03) as i32)?;
+            Ok(MessageSignature {
+                signature: RecoverableSignature::from_compact(&bytes[1..], recid)?,
+                compressed: ((bytes[0] - 27) & 0x04) != 0,
+            })
+        }
+
+        /// Attempt to recover a public key from the signature and the signed message.
+        ///
+        /// To get the message hash from a message, use [signed_msg_hash].
+        pub fn recover_pubkey<C: secp256k1::Verification>(
+            &self,
+            secp_ctx: &secp256k1::Secp256k1<C>,
+            msg_hash: sha256d::Hash
+        ) -> Result<PublicKey, secp256k1::Error> {
+            let msg = secp256k1::Message::from_slice(&msg_hash[..])?;
+            let pubkey = secp_ctx.recover(&msg, &self.signature)?;
+            Ok(PublicKey {
+                key: pubkey,
+                compressed: self.compressed,
+            })
+        }
+
+        /// Verify that the signature signs the message and was signed by the given address.
+        ///
+        /// To get the message hash from a message, use [signed_msg_hash].
+        pub fn is_signed_by_address<C: secp256k1::Verification>(
+            &self,
+            secp_ctx: &secp256k1::Secp256k1<C>,
+            address: &Address,
+            msg_hash: sha256d::Hash
+        ) -> Result<bool, secp256k1::Error> {
+            let pubkey = self.recover_pubkey(&secp_ctx, msg_hash)?;
+            Ok(match address.address_type() {
+                Some(AddressType::P2pkh) => {
+                    *address == Address::p2pkh(&pubkey, address.network)
+                }
+                Some(AddressType::P2sh) => false,
+                Some(AddressType::P2wpkh) => {
+                    // Only compressed pubkeys are allowed in p2wpkh.
+                    pubkey.compressed &&
+                        *address == Address::p2wpkh(&pubkey, address.network).expect("compressed pk")
+                }
```

Comment:
This is not aligned with bitcoin-core AFAICT 
https://github.com/bitcoin/bitcoin/blob/a12d9e5fd24a25bef476c10620317e43a5905754/src/util/message.cpp#L34

Reply:
Interesting. It looks like Core will *only* run `verifymessage` on p2pkh addresses. I think we should replicate this @stevenroose and implement BIP322 in a separate PR for new-style addresses.